From 611b5ea689f78a798cfd86aac92a0e4e4a41cafb Mon Sep 17 00:00:00 2001
From: Johannes Thumshirn <jthumshirn@suse.de>
Date: Thu, 8 Dec 2016 14:04:57 +0100
Subject: [PATCH] splice: introduce FMODE_SPLICE_READ and FMODE_SPLICE_WRITE
References: CVE-2016-9576, bsc#1013604
Patch-mainline: Submitted on 2016-12-09

Introduce FMODE_SPLICE_READ and FMODE_SPLICE_WRITE. These modes check
whether it is legal to read or write a file using splice. Both get
automatically set on regular files and are not checked when a 'struct
fileoperations' includes the splice_{read,write} methods.

Suggested-by: Linus Torvalds <torvalds@linux-foundation.org>
Cc: Al Viro <viro@zeniv.linux.org.uk>
Signed-off-by: Johannes Thumshirn <jthumshirn@suse.de>
---
 fs/open.c          |    4 ++++
 fs/splice.c        |    7 +++++++
 include/linux/fs.h |    5 +++++
 3 files changed, 16 insertions(+)

--- a/fs/open.c
+++ b/fs/open.c
@@ -701,6 +701,10 @@ static int do_dentry_open(struct file *f
 	if (S_ISREG(inode->i_mode))
 		f->f_mode |= FMODE_ATOMIC_POS;
 
+	if (S_ISREG(inode->i_mode))
+		f->f_mode |= FMODE_SPLICE_WRITE | FMODE_SPLICE_READ;
+
+
 	f->f_op = fops_get(inode->i_fop);
 	if (unlikely(WARN_ON(!f->f_op))) {
 		error = -ENODEV;
--- a/fs/splice.c
+++ b/fs/splice.c
@@ -621,6 +621,10 @@ ssize_t default_file_splice_read(struct
 		.spd_release = spd_release_page,
 	};
 
+	if (unlikely(!(in->f_mode & FMODE_SPLICE_READ)))
+		return -EINVAL;
+
+
 	if (splice_grow_spd(pipe, &spd))
 		return -ENOMEM;
 
@@ -1083,6 +1087,9 @@ static ssize_t default_file_splice_write
 {
 	ssize_t ret;
 
+	if (unlikely(!(out->f_mode & FMODE_SPLICE_WRITE)))
+		return -EINVAL;
+
 	ret = splice_from_pipe(pipe, out, ppos, len, flags, write_pipe_buf);
 	if (ret > 0)
 		*ppos += ret;
--- a/include/linux/fs.h
+++ b/include/linux/fs.h
@@ -136,6 +136,11 @@ typedef void (dio_iodone_t)(struct kiocb
 /* File was opened by fanotify and shouldn't generate fanotify events */
 #define FMODE_NONOTIFY		((__force fmode_t)0x1000000)
 
+/* File can be read using splice */
+#define FMODE_SPLICE_READ       ((__force fmode_t)0x8000000)
+/* File can be written using splice */
+#define FMODE_SPLICE_WRITE      ((__force fmode_t)0x10000000)
+
 /*
  * Flag for rw_copy_check_uvector and compat_rw_copy_check_uvector
  * that indicates that they should check the contents of the iovec are
